### ISSUE TITLE
Can't refresh token via Gitlab OIDC

### ISSUE THREAD

**Issue by wi-cuckoo** [goharbor/harbor#8069]
If you are reporting a problem, please make sure the following information are provided:

**Expected behavior and actual behavior:**

When I login via gitlab with OIDC, the login state will expire very soon, yeah, 120 second. I hope it can be longer.

**Steps to reproduce the problem:**

1. create a gitlab application, and select the scope for openid
2. then configure the harbor to login via OIDC
3. and auth works, but after 120s, it will return 401 of /api/users/current
4. check the core.log, will find it failed to refresh the token

**Versions:**
Please specify the versions of following systems.
- harbor version: [1.8.0]
- docker engine version: [18.06.1]
- docker-compose version: [1.22.0]

**Additional context:**
* **Harbor config files:** You can get them by packaging `harbor.cfg` and files in the same directory, including subdirectory.
* **Log files:** 
```
 109 Jun 17 20:28:02 172.19.0.1 core[3804654]: 2019-06-17T12:28:02Z [ERROR] [/common/api/base.go:69]: GET /api/users/current failed with error: {"code":401,"message":"UnAuthorize"}
 110 Jun 17 20:28:02 172.19.0.1 core[3804654]: 2019/06/17 12:28:02 #033[1;44m[D] [server.go:2774] |   10.115.21.44|#033[43m 401 #033[0m|   2.131508ms|   match|#033[44m GET     #033[0m /api/     users/current   r:/api/users/:id#033[0m
 111 Jun 17 20:28:02 172.19.0.1 core[3804654]: 2019-06-17T12:28:02Z [INFO] [/common/utils/oidc/secret.go:110]: Failed to verify ID Token, error: oidc: token is expired (Token Expiry: 2019-      06-17 12:27:03 +0000 UTC), refreshing...
 112 Jun 17 20:28:02 172.19.0.1 core[3804654]: 2019-06-17T12:28:02Z [ERROR] [/core/filter/security.go:509]: Failed to verify secret, error: failed to verify the secret: failed to get            id_token from refresh response
```

**Comment by reasonerjt** [comment]
This maybe due to misconfiguration

You need to double check if Gitlab returns refresh token or not and what's the response of the refresh request.

**Comment by wi-cuckoo** [comment]
@reasonerjt yeah，the response has nothing, I logged it and the raw field is nil. I'm not sure why gitlab doesn't work as expect. I just do as gitlab doc tells. so confusing

**Comment by reasonerjt** [comment]
Did you setup the `scope` correctly in Harbor's configuration tab?

**Comment by wi-cuckoo** [comment]
@reasonerjt the `scope` just set as `openid,profile,email,api`, and in gitlab checked such four checkboxes

**Comment by Mierpo** [comment]
What is the correct configuration for this in Harbor? We're also running into this issue. We're also using Gitlab. Our scope is slightly different though: read_user,openid,profile,email. Harbor version 1.8.1.

**Comment by reasonerjt** [comment]
Well, conventionally you need the `offline_access`  to make sure OIDC provider returns the refresh token.  
But that's provider specific setting, please check the manual of the OIDC provider you are integrating with.

**Comment by wi-cuckoo** [comment]
@reasonerjt  yeah, I noticed the `offlie_access` scope，but the gitlab doesn't support it, that's confusing. I don't know how to solve this problem, except for modifying source code in harbor to extend the token expiration manually

**Comment by reasonerjt** [comment]
@wi-cuckoo I checked the doc:
https://docs.gitlab.com/ee/api/oauth2.html
Could you verify using `curl` command to see if your gitlab returns `refresh_token` or not?
Additionally you should not modify Harbor's code to extend the token, because the token is returned by gitlab?

**Comment by lixiaofei123** [comment]
the same question，login is expire after about 120 seconds,and i have to login harbor again and get new password to push image

**Comment by jsimomaa** [comment]
@reasonerjt I did some testing with our GitLab and here is the results of `curl`:

```bash
$ curl -d "client_id=01969f372ccc5c7902a920ad6d5e90f656c138bb549ab4001c9c7de829fa5b97&client_secret=e08b16273072aad7d868eb633daf80084d87dddc1bf1e3f454119357a94f8687&code=f8c1167af8d4cdbd6b480864c27e6c79b9c59c84dcb1cef8a39f289d3fa9c153&grant_type=authorization_code&redirect_uri=http://example.com" https://gitlab.example.com/oauth/token
{"access_token":"667281847bd721486c5bed6510213519f83d90bd3186a1b5cec27098297906e3","token_type":"bearer","refresh_token":"d43df613353135e8768d36179d43db7b657da7dd345081f5a3fddd13561747f8","scope":"api","created_at":1562314777}
```

So indeed a `refresh_token` is returned. Then I found [this issue](https://gitlab.com/gitlab-org/gitlab-ce/issues/44485) that mentioned that `id_token` might or might not be present when invoking a refresh request. I tested this via:

```bash
$ curl -d "grant_type=refresh_token&refresh_token=6b4e887b6fb0bb00061732ef6a868248b85a7f78425a7f422447c905353edb97" https://gitlab.example.com/oauth/token
{"access_token":"7fda4f768063d14a1ecd0e396416ae036ac34ea4439df210b4fa5978afd6de66","token_type":"bearer","refresh_token":"b44dc4346df6c4dd4949071b6f73dc089b675cfc446826cfa734f117a8ff1c60","scope":"api","created_at":1562315122}
```

And indeed it seems that the `id_token` is somehow missing from the response of the refresh request.

From the issue above:

> Digging further I read in the spec
( http://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse ): "the response body is the Token Response of Section 3.1.3.3 except that it might **not** contain an `id_token`."

**Comment by jsimomaa** [comment]
Seems like I had the wrong scope `api` when I did the initial authorization. With scope `openid` the `curl` invocations become:

```bash
$ curl -d "client_id=01969f372ccc5c7902a920ad6d5e90f656c138bb549ab4001c9c7de829fa5b97&client_secret=e08b16273072aad7d868eb633daf80084d87dddc1bf1e3f454119357a94f8687&code=b59f57f66f22f8c0947129dee1825fb136c51c59f2bb84f7877c53e9a3453db5&grant_type=authorization_code&redirect_uri=http://example.com" https://gitlab.example.com/oauth/token
{"access_token":"bc6dfa4b73a59f9a827bf7f7b1d02e5c3a4038dcd24f47ebc7935726075f1618","token_type":"bearer","refresh_token":"e2d746b8fcb0c139faec6e6363a112749d989b4a07eed5e6be28b818a5290971","scope":"openid","created_at":1562326124,"id_token":"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjhaTWM3TzhNQ2MzNTNWMHJGQzhhOVA1NS1RMFNLYzd3a1pUbGJiRE5tckkifQ.eyJpc3MiOiJodHRwczovL2dpdGxhYi5zaW11cGVkaWEuY29tIiwic3ViIjoiMiIsImF1ZCI6IjAxOTY5ZjM3MmNjYzVjNzkwMmE5MjBhZDZkNWU5MGY2NTZjMTM4YmI1NDlhYjQwMDFjOWM3ZGU4MjlmYTViOTciLCJleHAiOjE1NjIzNTMwMDgsImlhdCI6MTU2MjM1Mjg4OCwiYXV0aF90aW1lIjoxNTYyMzUwNjk5LCJzdWJfbGVnYWN5IjoiYTdiMDRjYWEyNWNjNjUyMTA5NzViZWExNDViMTcwNTI1NzYwZTUyMjRjOWZjMjhiNTQ1YTQ1YmRmMDFlNWRlYSJ9.BcvW6jxl5BXOGTY_EW08MsMTMVTOE_S1kHfVeTiPpm3IRrB6xiqcPpPwv62ykr0XorBi6t2q4wWKYCgcrw1NU7hMfYu-HY2JMByRxpIXal7ZH9xsHLv8cd6jQIye0mxe2-QkkBfp-G9vW7txmpsgDPT0ITF9Bl5cNsmpi5iyiuZtEMiSy7Y6Iv04pZ-74s67bQOmabl_U83GCxtec9h5Mlwm6lQcLbahxXlmIjApWzoB4MStMNGJlMWMrYmfbCj2YCUZwGmyMcniQ6rsHWHJt_AbGAKM9gAkpjnWkJUo2x8RmwK-NUC7-8c1kB7xYmUTGS7IHjdn_JQze0AP4ZLiogxrkzR_QpDC9qWbiAHMglCUHq7iVeTcysabdmoo-LzIrV4VcvsdfsTJA_2HzW-y9OahrWnCrJociIb78d7e8bvJ7OoTzOUg6M40UdFMYtDBK8uYdDw2DSZXZXAoK3dIzqH6p2BHciJ6quFbi08TCGmhySkWFoZhZ25dmN-RwrMvmH1W1CqXO5yi_enVJk6DXM1P46eGWiXs_EEEi2TuI6oYXhH-zhW1Ra4c5b7OxMb8VewJ1QGUOOKwgRy9wc1B0o4fumCFkDfioXbtG0A1GDVknzi6Q6MQqw9tfD5EW69f09Pr4E9bHIirKyyIKrMez7Rn6VJMJX3ah1nHTOLERbQ"}
```

And refresh request:

```bash
$ curl -d "grant_type=refresh_token&refresh_token=e2d746b8fcb0c139faec6e6363a112749d989b4a07eed5e6be28b818a5290971" https://gitlab.example.com/oauth/token
{"access_token":"dbcf57e02c0694172a60c5fcfa9aadd4a94590129ebe1817b0e6ba40fa17ec99","token_type":"bearer","refresh_token":"5c65d0e068ad6a8d5fcaa87acf1253736c9eb1813658c09980fc195067c3d2fc","scope":"openid","created_at":1562352926,"id_token":"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjhaTWM3TzhNQ2MzNTNWMHJGQzhhOVA1NS1RMFNLYzd3a1pUbGJiRE5tckkifQ.eyJpc3MiOiJodHRwczovL2dpdGxhYi5zaW11cGVkaWEuY29tIiwic3ViIjoiMiIsImF1ZCI6IjAxOTY5ZjM3MmNjYzVjNzkwMmE5MjBhZDZkNWU5MGY2NTZjMTM4YmI1NDlhYjQwMDFjOWM3ZGU4MjlmYTViOTciLCJleHAiOjE1NjIzNTMwNDYsImlhdCI6MTU2MjM1MjkyNiwiYXV0aF90aW1lIjoxNTYyMzUwNjk5LCJzdWJfbGVnYWN5IjoiYTdiMDRjYWEyNWNjNjUyMTA5NzViZWExNDViMTcwNTI1NzYwZTUyMjRjOWZjMjhiNTQ1YTQ1YmRmMDFlNWRlYSJ9.FA42AIR7Ab6wsgWAY25zmhO-kEgmduqpG_gemaYrP6dSsNneXiWBPTyJVOLH1ygwZXlUzEl_8KckotG8uJGry0gEEIwpvo_x9J2DNkrK7JDwQDRGCsa6pHl_ziL2ApCMzqkWRwojT1KxhEVhhy4V3voGhGgx1oi95GWEh9AyTUx6uypB5QG72Eehjuf6dHYKW2oWhPFI_4t46-pl2jZqrv9kP2MNe0-iBrCT_49KmYNwoVLD1TkDOoLFvutDwBg9yZecWbsiE7DH-rxyK7RC0SrE6yjNFht4AWZkhMDbYEeSYrz4o8PU9fEHjRx7UZpDFIMd9_ahqnpOP2VjiepLMsHYyW5Mi6jDv7I6sNNaTmPcIBAfKY_a2pvQj5SrC3_atwnO6TmExhtyMnY77BVxT6qhDWoytSTlTdgRDA6TvE55RZdWysLIByuAVduu2jt73i-zWCY1dgbLW_zIydcUeT7bEPxm4JtDPrZ7BreweLmNz_7SY_Sr7VlFevL_LYR2icPnaSK3NDKNAO1Rp6sKk6vShdeRZY_x4Z6aSGnHXKcx6dC9Agpvbmyz942ennViuQGQ1ImNoCZKFiKTv2C97aQSaMiwSMhCChD487FiTWGUCAA0BB-ydPgjIHizoOcn1sojO9EopWanPEQRnXZCIvQYn2e1K8lHHRmq3takdBQ"}
```
So actually the `ìd_token` is present on the refresh response.

I did some more investigation and here are the relevant log entries for both `/var/log/nginx/gitlab_access.log` and `/var/log/harbor/core.log`:

`/var/log/nginx/gitlab_access.log`:
```bash
172.30.4.134 - - [05/Jul/2019:22:11:48 +0300] "GET /.well-known/openid-configuration HTTP/2.0" 200 1192 "-" "Go-http-client/2.0"
172.30.4.134 - - [05/Jul/2019:22:11:48 +0300] "GET /oauth/authorize?client_id=993d675b4dae344ff881f34e1c0c1f42e0272b899f6e1f3b81522e9778bd083d&redirect_uri=http%3A%2F%2Fharbor.testing%2Fc%2Foidc%2Fcallback&response_type=code&scope=openid&state=knwiw6q0qx02lzdm08hlof4hh11dlzc4 HTTP/2.0" 302 606 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)
172.30.4.134 - 993d675b4dae344ff881f34e1c0c1f42e0272b899f6e1f3b81522e9778bd083d [05/Jul/2019:22:11:48 +0300] "POST /oauth/token HTTP/2.0" 200 1637 "-" "Go-http-client/2.0"
172.30.4.134 - - [05/Jul/2019:22:11:48 +0300] "GET /oauth/discovery/keys HTTP/2.0" 200 1139 "-" "Go-http-client/2.0"
# at this point the token has expired and refresh request is expected
172.30.4.134 - - [05/Jul/2019:22:15:34 +0300] "GET /.well-known/openid-configuration HTTP/2.0" 200 1192 "-" "Go-http-client/2.0"
172.30.4.134 - - [05/Jul/2019:22:15:34 +0300] "GET /.well-known/openid-configuration HTTP/2.0" 200 1192 "-" "Go-http-client/2.0"
```

Output from `docker-compose up`:

```bash
harbor-core    | 2019/07/05 19:11:46 [D] [server.go:2774] |     172.30.0.1| 200 |     4.3459ms|   match| GET      /c/log_out   r:/c/log_out
nginx          | 172.30.0.1 - "GET /c/log_out HTTP/1.1" 200 0 "http://harbor.testing/harbor/configs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36" 0.005 0.010 .
harbor-core    | 2019/07/05 19:11:46 [D] [server.go:2774] |     172.30.0.1| 200 |     5.3445ms|   match| GET      /api/systeminfo   r:/api/systeminfo
nginx          | 172.30.0.1 - "GET /api/systeminfo HTTP/1.1" 200 242 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36" 0.006 0.000 .
harbor-core    | 2019/07/05 19:11:48 [D] [server.go:2774] |     172.30.0.1| 302 |    175.132ms|   match| GET      /c/oidc/login   r:/c/oidc/login
nginx          | 172.30.0.1 - "GET /c/oidc/login HTTP/1.1" 302 292 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36" 0.176 0.180 .
harbor-core    | 2019/07/05 19:11:48 [D] [server.go:2774] |     172.30.0.1| 302 |    67.2039ms|   match| GET      /c/oidc/callback   r:/c/oidc/callback
nginx          | 172.30.0.1 - "GET /c/oidc/callback?code=bc4abeebfba91426d198ff7a3d58363567c9d96eaf382ff6f1faab56924a6f69&state=knwiw6q0qx02lzdm08hlof4hh11dlzc4 HTTP/1.1" 302 46 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36" 0.068 0.060 .
nginx          | 172.30.0.1 - "GET /oidc-onboard?username= HTTP/1.1" 200 693 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36" 0.000 0.000 .
harbor-portal  | 172.19.0.10 - - [05/Jul/2019:19:11:48 +0000] "GET /oidc-onboard?username= HTTP/1.1" 200 693 "http://harbor.testing/harbor/sign-in?signout=true&redirect_url=%2Fharbor%2Fconfigs" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36"

# then at the time of refresh request

harbor-core    | 2019-07-05T19:15:33Z [INFO] [/common/utils/oidc/secret.go:110]: Failed to verify ID Token, error: oidc: token is expired (Token Expiry: 2019-07-05 19:13:48 +0000 UTC), refreshing...
harbor-core    | 2019-07-05T19:15:33Z [ERROR] [/core/filter/security.go:511]: Failed to verify secret, error: failed to verify the secret: failed to get id_token from refresh response
```

I would have suspected another `POST /oauth/token` in the `gitlab_access.log` for refresh?

**Comment by reasonerjt** [comment]
@jsimomaa 
I don't have a quick idea why harbor is not sending the refresh request.
Sorry for the late response, could you switch the log level to `debug` and provide the output?

I'll see if I can setup gitlab as OIDC provider, but that has to be a few days later.

**Comment by mkjpryor-stfc** [comment]
I am also affected by this.

**Comment by reasonerjt** [comment]
@jsimomaa 
After re-reading the comments and #9267 I realize it's maybe correct that we do not expect id token in the refresh response.
thanks for reporting this.

**Comment by reasonerjt** [comment]
fixed in #9481 and #9435
